cosmpy/clients/ledger.py

```diff
@@ -0,0 +1,1029 @@
+                )
+
+                tx = self.generate_tx(
+                    [msg],
+                    [sender_crypto.get_address()],
+                    [sender_crypto.get_pubkey_as_bytes()],
+                    gas_limit=gas,
+                )
+                self.sign_tx(sender_crypto, tx)
+
+                res = self.broadcast_tx(tx)
+                code_id = self.get_code_id(res)
+            except BroadcastException as e:
+                # Failure due to wrong sequence, signature, etc.
+                last_exception = e
+                _logger.warning(
+                    "Failed to deploy contract code due BroadcastException: {%s}", e
+                )
+                self._sleep(self.msg_failed_retry_interval)
+                attempt += 1
+                continue
+
+        if code_id is None or res is None:  # pragma: nocover
+            raise BroadcastException(
+                f"Failed to deploy contract code after multiple attempts: {last_exception}"
+            )
+
+        return code_id, MessageToDict(res)
+
+    @staticmethod
+    def get_code_id(response: GetTxResponse) -> int:
+        """
+        Get code id from store code transaction response
+        :param response: Response of store code transaction
+        :return: integer code_id
+        """
+        raw_log = json.loads(response.tx_response.raw_log)
+        assert raw_log[0]["events"][1]["attributes"][0]["key"] == "code_id"
+        return int(raw_log[0]["events"][1]["attributes"][0]["value"])
+
+    @staticmethod
+    def get_contract_address(response: GetTxResponse) -> str:
+        """
+        Get contract address from instantiate msg response
+        :param response: Response of MsgInstantiateContract transaction
+        :return: contract address string
+        """
+        raw_log = json.loads(response.tx_response.raw_log)
+        assert raw_log[0]["events"][0]["attributes"][0]["key"] == "_contract_address"
+        return str(raw_log[0]["events"][0]["attributes"][0]["value"])
+
+    def instantiate_contract(
+        self,
+        sender_crypto: CosmosCrypto,
+        code_id: int,
+        init_msg: JSONLike,
+        label: str,
+        gas: int = DEFAULT_GAS_LIMIT,
+    ) -> Tuple[str, JSONLike]:
+        """
+        Send init contract message
+
+        :param sender_crypto: Deployer crypto to sign init message
+        :param code_id: ID of binary code stored on chain
+        :param init_msg: Init message in json format
+        :param label: Label of current instance of contract
+        :param gas: Gas limit
+
+        :return: Contract address string, transaction response
+
+        :raises BroadcastException: When communication with node fails.
+        """
+        elapsed_time = 0
+        res: Optional[GetTxResponse] = None
+        contract_address: Optional[str] = None
+        last_exception: Optional[Exception] = None
+        while contract_address is None and elapsed_time < self.n_total_msg_retries:
```

Comment:
Having this loop with retries and catching BroadcastException as a context manager will reduce code duplication

Reply:
Will be done in next PR